[WIP] Add Calendar layout

Nitwel opened the pull request:
![grafik](https://user-images.githubusercontent.com/22577866/91359177-54dc9480-e7f4-11ea-8b05-f35b2652b486.png)

Adds the calendar layout to the next gen version of directus. 🚀 

<details>
  <summary>More Images</summary>
Week view

![grafik](https://user-images.githubusercontent.com/22577866/91360052-07612700-e7f6-11ea-8022-8de8ac7020a6.png)

Agenda view

![grafik](https://user-images.githubusercontent.com/22577866/91360163-32e41180-e7f6-11ea-9c91-b0bb573be671.png)

</details>

**Features**
- Normal calendar view.
- Time relative weekly view and single day view.
- Agenda view (Timeline)
- Support for Datetime, Date, Date&Time,
- Color support

**Todo**
- [ ] Add time axis to day view (if  wanted)
- [ ] Clean up code
- [ ] Scrolling interactions
- [ ] Click interactions (Click on day...)
- [ ] Year view (maybe?)
- [ ] Proper way to display all events in the month view
- [ ] Support status colors (Dont know how to do that yet properly)
- [ ] Get the collection header smaller?
- [ ] Fix layout height
- [ ] Add on specific date button

rijkvanzanten commented on app/src/layouts/calendar/calendar.vue:
```diff
@@ -0,0 +1,567 @@
+
+			return { datetime, date, time, title, color, isDatetime, isAgenda, animations };
+
+			function createViewOption<T>(key: keyof ViewOptions, defaultValue: any) {
+				return computed<T>({
+					get() {
+						return _viewOptions.value?.[key] !== undefined ? _viewOptions.value?.[key] : defaultValue;
+					},
+					set(newValue: T) {
+						_viewOptions.value = {
+							..._viewOptions.value,
+							[key]: newValue,
+						};
+					},
+				});
+			}
```
Should we make this a util function? It's re-used between tabular/cards/calendar right?

Nitwel replied:
I guess that makes sense!

rijkvanzanten commented on app/src/layouts/calendar/components/agenda.vue:
```diff
@@ -0,0 +1,231 @@
+						<div class="line"><div class="dot"></div></div>
+						<div class="content" @click="newItem(i)">
+							<v-icon name="add"></v-icon>
+							add new Item
```
Use translations

rijkvanzanten commented on app/src/layouts/calendar/components/agenda.vue:
```diff
@@ -0,0 +1,231 @@
+			</div>
+		</div>
+		<div v-else class="no-events">
+			<span>There are no events in between</span>
```
use translations

rijkvanzanten commented on app/src/layouts/calendar/components/agenda.vue:
```diff
@@ -0,0 +1,231 @@
+		</div>
+		<div v-else class="no-events">
+			<span>There are no events in between</span>
+			<span>{{ interval.getStart().toLocaleDateString() }} and {{ interval.getEnd().toLocaleDateString() }}</span>
```
use translations

rijkvanzanten commented on app/src/layouts/calendar/components/agenda.vue:
```diff
@@ -0,0 +1,231 @@
+						</div>
+					</div>
+					<div class="event add">
+						<div class="line"><div class="dot"></div></div>
```
prefer self-closing for empty elements

```suggestion
						<div class="line"><div class="dot"/></div>
```

rijkvanzanten commented on app/src/layouts/calendar/components/agenda.vue:
```diff
@@ -0,0 +1,231 @@
+			router.push(item.__link__);
+		}
+
+		function newItem(index: number) {
```
Should be disabled in select mode

rijkvanzanten commented on app/src/layouts/calendar/components/agenda.vue:
```diff
@@ -0,0 +1,231 @@
+				}
+
+				if (time) {
+					item.__time__ = time;
```
I don't know how hard this is / if it's possible, but can we use an object structure instead of the `__` for the system data? Like

```js
const event = {
  meta: {
    time: 'xxx'
  },
  item: { ... }
};
```

instead of

```js
const event = {
   ...item,
   __time__: 'xxx'
};
```

rijkvanzanten commented on app/src/layouts/calendar/components/agenda.vue:
```diff
@@ -0,0 +1,231 @@
+					}
+
+					&:last-child {
+						.line {
+							background-color: transparent;
+						}
```
Can we flatten out the CSS here a bit?

Vue already does the scoping, so we can keep the SCSS fairly flat

benhaynes commented on app/src/layouts/calendar/calendar.vue:
```diff
@@ -0,0 +1,567 @@
+<template>
+	<div class="calendar">
+		<portal to="layout-options">
+			<div class="layout-option">
+				<div class="option-label">{{ $t('layouts.calendar.animations') }}</div>
+				<v-switch
+					v-model="animations"
+					:label="$t(`layouts.calendar.animations-${animations ? 'enabled' : 'disabled'}`)"
+				></v-switch>
+			</div>
```
Yes, agree.

benhaynes commented on app/src/layouts/calendar/calendar.vue:
```diff
@@ -0,0 +1,567 @@
+			return options;
+		});
+
+		function onScroll(event: WheelEvent) {
+			if (viewType.value != Interval.Type.MONTH) return;
+			if (event.deltaY > 0) forwards();
+			else backwards();
+		}
```
Yeah, pulling from scroll events can be all over the place on different devices/OSs... gets annoying fast when you swipe through 45 years with one swipe.

rijkvanzanten commented:
My fears are true. The scroll to forward/backward doesn't work on macs i'm afraid:

https://cln.sh/KORXSk

Nitwel commented:
Well, it is the same as in v8 so either we ignored that in v8 or idk.

rijkvanzanten commented:
Random thoughts/questions while using this:

* The scroll handlers don't play nicely with trackpads / touchscreens, lets remove those.

* Can we make the years / months / week view overflow scroll w/ virtualization on the days instead of separate "sections" that toggle back and forth? (combined with scroll snap sections)
  * This'll also solve for trackpad use
  * This'll also solve for the height: 100% hack 

* Can we default the layout options smartly to an available date/time/datetime field?

* Lets add a fallback / error state for when the layout options aren't setup, and/or there aren't any datetime fields available

* Timestamp fields don't show up:

![CleanShot 2020-09-17 at 14 00 03](https://user-images.githubusercontent.com/9141017/93509706-1f583200-f8ee-11ea-80af-f531cdb852d8.png)

* Lets use the default form styling in here:

![CleanShot 2020-09-17 at 14 00 30](https://user-images.githubusercontent.com/9141017/93509737-297a3080-f8ee-11ea-8e35-2fe834b60687.png)

* Can we change `title` for a display template? That way it's inline with the options for the `cards` layout

Nitwel commented:
> The scroll handlers don't play nicely with trackpads / touchscreens, lets remove those.

Removed it.

> Can we make the years / months / week view overflow scroll w/ virtualization on the days instead of separate "sections" that toggle back and forth? (combined with scroll snap sections)

As far as I understood you right here, I don'think thats a good idea here for several reasons:
- All views would have to render at the same time and that would be aweful
- How'd you handle week and day view? I don't think we wan't a double scrollbar. 😄 
  - And don't say scroll use 1 scrollbar and have the week larger than 100% cause that would cause problems with scroll snap 😉 

> Can we default the layout options smartly to an available date/time/datetime field?

Yes we could do that but then we should implement this for all layouts not just this one.

> Lets add a fallback / error state for when the layout options aren't setup, and/or there aren't any datetime fields available

Hmm, I wouldn't remove the whole calendar and have an error message there but we could do this either in the sidebar or in the header i think?

> Timestamp fields don't show up:

Yea, I added that now!

> Lets use the default form styling in here

I'm not exactly sure what you mean here. At some point in time this all looked right but now it doesn't anymore and I don't know what you or someone else might have changed here. 😐 (All other layouts have non spaced options now too)

> Can we change title for a display template? That way it's inline with the options for the cards layout

Sounds good, I'll do that now.

danieljwestman commented:
How is this calendar coming along in v9? Looking forward seeing it as it was a great feature/UI in v8.

Nitwel commented:
It's mostly done just needs some review time but @rijkvanzanten has a lot of other stuff on his list to do first.
When that's more empty he will review and merge it.

danieljwestman commented:
@Nitwel Lovely!

rijkvanzanten commented:
There's a lot of moving pieces in this one, so i wanna make sure it's nice and neat 🙂
